crates/ra_hir/src/ty.rs

```diff
@@ -989,19 +1000,24 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
     /// If `ty` is a type variable with known type, returns that type;
     /// otherwise, return ty.
     fn resolve_ty_shallow<'b>(&mut self, ty: &'b Ty) -> Cow<'b, Ty> {
-        match ty {
-            Ty::Infer(tv) => {
-                let inner = tv.to_inner();
-                match self.var_unification_table.probe_value(inner).known() {
-                    Some(known_ty) => {
-                        // The known_ty can't be a type var itself
-                        Cow::Owned(known_ty.clone())
+        let mut ty = Cow::Borrowed(ty);
+        for _ in 0..3 {
```

Comment:
This might deserve a comment (I'm assuming it's arbitrary). Maybe a message if we fail to resolve in that time?

Reply:
A mark might be useful here as well!

Reply:
It's not quite arbitrary -- we have normal, int and float type variables, and they might be nested, but each only once (even an int variable resolving to a float variable shouldn't happen, but better to make sure).

I added a mark and comment.